File: code/backend/Odin/src/main/kotlin/pt/isel/odin/service/interfaces/UserService.kt
package pt.isel.odin.service.interfaces

import pt.isel.odin.controller.dto.user.UserRequest
import pt.isel.odin.model.User

/**
 * Service for users.
 */
interface UserService {

    /**
     * Gets a user by its id.
     *
     * @param id the user id.
     */
    fun getById(id: String): User

    /**
     * Gets all users.
     */
    fun getAll(): List<User>

    /**
     * Saves or updates a user.
     *
     * @param userRequest the user to save.
     */
    fun save(userRequest: UserRequest): User

    /**
     * Deletes a user by its id.
     *
     * @param id the user id.
     */
    fun delete(id: String)
}

File: code/frontend/Tyr/src/services/UserService.ts
import {makeApiRequest} from "../axios/apiRequest";
import {User} from "../model/User";

/**
 * Service to manage users
 */
export class UserService {

    static basePath = '/users';

    static async getSession(): Promise<User> {
        return makeApiRequest('get', `${(UserService.basePath)}/session`);
    }

    static async getById(id: string): Promise<User> {
        return makeApiRequest('get', `${(UserService.basePath)}/${id}`);
    }

    static async getAll(): Promise<User[]> {
        return makeApiRequest('get', UserService.basePath);
    }

    static async save(userRequest: User): Promise<User> {
        return makeApiRequest('post', `${(UserService.basePath)}/save`, userRequest);
    }

    static async update(userRequest: User): Promise<User> {
        return makeApiRequest('put', `${(UserService.basePath)}/update`, userRequest);
    }

    static async delete(id: number): Promise<void> {
        return makeApiRequest('delete', `${(UserService.basePath)}/${id}`);
    }
}


